dask_distance/__init__.py

```diff
@@ -93,6 +93,35 @@ def cdist(XA, XB, metric="euclidean"):
     return result
 
 
+def pdist(X, metric="euclidean"):
+    """
+    Finds the pairwise condensed distance matrix using the metric.
+
+    Args:
+        X:          2-D array of points
+        metric:     string or callable
+
+    Returns:
+        array:      condensed distance between each pair
+
+    Note:
+        Tries to avoid redundant computations as much as possible.
+        However this is limited in its ability to do this based on
+        the chunk size of X (particularly along the first dimension).
+        Smaller chunks will increase savings though there may be
+        other tradeoffs.
+    """
+
+    result = cdist(X, X, metric)
+
+    result = dask.array.triu(result, 1)
```

Comment:
At first this may seem redundant with the mask applied below. However this is not the case. Instead this is what makes `pdist` optimized over `cdist` for pairs.

Using `triu` makes sure that any chunks completely on or below the central diagonal are replaced with `zeros`. This detaches anything that was connected to those chunks previously. So those chunks won't be computed at all. There is admittedly some cost to creating these arrays of zeros, but this cost should be amortized over the metric computations and should be much less compared to computing the metric.